Provide a RetryOnConflict helper for client libraries

smarterclayton opened the pull request:
Used like:

```
var pod *api.Pod
err := client.RetryOnConflict(client.DefaultBackoff, func() (err error) {
  pod, err = c.Pods("mynamespace").UpdateStatus(update)
  return
})
// err may be conflict
```

Triggered by openshift/origin#5301 and other client code scattered around

smarterclayton committed:
Provide a RetryOnConflict helper for client libraries

Used like:

    var pod *api.Pod
    err := client.RetryOnConflict(client.DefaultBackoff, func() (err error) {
      pod, err = c.Pods("mynamespace").UpdateStatus(update)
      return
    })
    // err may be conflict

bgrant0607 commented:
Assigned to @lavalamp to review or delegate.

lavalamp commented on pkg/client/unversioned/conditions.go:
```diff
@@ -17,11 +17,46 @@ limitations under the License.
+// RetryConflict executes the provided function repeatedly, retrying if the server returns a conflicting
+// write. Callers should preserve previous executions if they wish to retry changes. It performs an
+// exponential backoff.
+// TODO: Make Backoff an interface?
```
Add the PR example comment here.

lavalamp commented on pkg/client/unversioned/conditions.go:
```diff
@@ -17,11 +17,46 @@ limitations under the License.
+// TODO: Make Backoff an interface?
+func RetryOnConflict(backoff wait.Backoff, fn func() error) error {
+	var lastConflictErr error
+	err := wait.ExponentialBackoff(backoff, func() (bool, error) {
```
Is exponential backoff correct for this? This isn't like TCP where the same two packets collide over and over. If you got a conflict, that means someone else succeeded.

smarterclayton replied:
I was torn on this (and the default backoff may not be the right numbers).  A common case we're wanting to use this in is where multiple clients are competing to update the same entry.  We have a few controllers that are performing idempotent actions on the same resource, but we also have actions driven by clients where the initial action triggers a flurry of updates.

For example, our deployment controller (in OpenShift) deals with the deployment trigger controller updating the same resources.  That's a case where straight up retry is appropriate.  The second case is creating a deployment from the CLI - the first action (create) results in a bunch of controllers acting on it to create status.  In that case, backoff helps reduce the conflict.

I can add a `DefaultRetry` that has Factor 1.0.  I'll also find one of the instances in the code and replace it.

lavalamp commented on pkg/util/wait/wait.go:
```diff
@@ -40,6 +40,37 @@ var ErrWaitTimeout = errors.New("timed out waiting for the condition")
+}
+
+// ExponentialBackoff repeats a condition check up to steps times, increasing the wait
+// by multipling the previous duration by maxFactor. If jitter is greater than zero,
```
s/max//

lavalamp commented:
Minor comments.

smarterclayton committed:
Update service account tokens controller to use client.RetryOnConflict

smarterclayton committed:
Fix unrelated log level error

smarterclayton commented:
@deads2k given the complaints, we can potentially factor this differently.

lavalamp commented on pkg/client/unversioned/conditions.go:
```diff
@@ -17,11 +17,69 @@ limitations under the License.
+	Jitter:   0.1,
+}
+
+// DefaultBackoff is the recommended backoff for a conflict where a client
```
You seem to be making a distinction between Backoff and Retry, is it a common one?

smarterclayton replied:
I have historically differentiated - retry is the expectation that the conflict is random, backoff is intended where the caller  expects a flurry of modifications post creation (pods -> backoff, because of scheduling and kubelet, while retry is unrelated modification).

lavalamp commented on pkg/client/unversioned/conditions.go:
```diff
@@ -17,11 +17,69 @@ limitations under the License.
+var DefaultBackoff = wait.Backoff{
+	Steps:    4,
+	Duration: 10 * time.Millisecond,
+	Factor:   5.0,
```
The assumption that would make this make sense is that a controller that just made a change is less likely to make another change in the more distant future, but I don't see why that assumption should be true.

smarterclayton replied:
In practice a lot of controllers do a flurry of updates right after a
resource is created.  We see this on deployments, builds, pods, replication
controllers, and PVCs.  And service accounts.

On Oct 28, 2015, at 7:16 PM, Daniel Smith <notifications@github.com> wrote:

In pkg/client/unversioned/conditions.go
<https://github.com/kubernetes/kubernetes/pull/16067#discussion_r43331509>:

> +// DefaultRetry is the recommended retry for a conflict where multiple clients
> +// are making changes to the same resource.
> +var DefaultRetry = wait.Backoff{
> +	Steps:    5,
> +	Duration: 10 * time.Millisecond,
> +	Factor:   1.0,
> +	Jitter:   0.1,
> +}
> +
> +// DefaultBackoff is the recommended backoff for a conflict where a client
> +// may be attempting to make an unrelated modification to a resource under
> +// active management by one or more controllers.
> +var DefaultBackoff = wait.Backoff{
> +	Steps:    4,
> +	Duration: 10 * time.Millisecond,
> +	Factor:   5.0,

The assumption that would make this make sense is that a controller that
just made a change is less likely to make another change in the more
distant future, but I don't see why that assumption should be true.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/16067/files#r43331509>.

lavalamp commented on pkg/client/unversioned/conditions.go:
```diff
@@ -17,11 +17,69 @@ limitations under the License.
+//     ...
+//
+// TODO: Make Backoff an interface?
+func RetryOnConflict(backoff wait.Backoff, fn func() error) error {
```
Minor, but I would consider making Retry a method of Backoff, so that users write `client.DefaultBackoff.RetryOnConflict(fn)` instead of `client.RetryOnConflict(client.DefaultBackoff, fn)`.

Hm, I guess that doesn't work when backoff is from another package.

smarterclayton replied:
The cockroachdb guys had a good set of utilities for this, I need to review
what they did and crib more closely.  Also, David noticed and I agreed that
the back off/poll callback function is more useful than retry on conflict.
I need to see whether I can find a happier refactoring.  Hold slightly.

On Oct 28, 2015, at 7:18 PM, Daniel Smith <notifications@github.com> wrote:

In pkg/client/unversioned/conditions.go
<https://github.com/kubernetes/kubernetes/pull/16067#discussion_r43331645>:

> +// write. Callers should preserve previous executions if they wish to retry changes. It performs an
> +// exponential backoff.
> +//
> +//     var pod *api.Pod
> +//     err := RetryOnConflict(DefaultBackoff, func() (err error) {
> +//       pod, err = c.Pods("mynamespace").UpdateStatus(podStatus)
> +//       return
> +//     })
> +//     if err != nil {
> +//       // may be conflict if max retries were hit
> +//       return err
> +//     }
> +//     ...
> +//
> +// TODO: Make Backoff an interface?
> +func RetryOnConflict(backoff wait.Backoff, fn func() error) error {

Minor, but I would consider making Retry a method of Backoff, so that users
write client.DefaultBackoff.RetryOnConflict(fn) instead of
client.RetryOnConflict(client.DefaultBackoff,
fn).

Hm, I guess that doesn't work when backoff is from another package.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/16067/files#r43331645>.

jayunit100 commented on pkg/util/wait/wait.go:
```diff
@@ -40,6 +40,37 @@ var ErrWaitTimeout = errors.New("timed out waiting for the condition")
+// a random amount of each duration is added (between duration and duration*(1+jitter)).
+// If the condition never returns true, ErrWaitTimeout is returned. All other errors
+// terminate immediately.
+func ExponentialBackoff(backoff Backoff, condition ConditionFunc) error {
```
am  now discussing the connection to  backoff.go with @smarterclayton

lavalamp commented:
Is this PR still live?

eparis commented:
@k8s-bot test this [triggering because no jenkins unit test present and shippable not good enough any more]

smarterclayton commented:
Yes.  I decided against changing the API after using it a bit more.  It has
been useful as is.  Ready for review.

On Dec 1, 2015, at 6:31 PM, Daniel Smith <notifications@github.com> wrote:

Is this PR still live?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/16067#issuecomment-161130623>
.

krousey commented on pkg/controller/namespace/namespace_controller.go:
```diff
@@ -232,26 +233,26 @@ func deleteAllContent(kubeClient client.Interface, versions *unversioned.APIVers
 		}
-	}
+		return err
+	})
```
Shouldn't this function be replaced by the client function now instead of just another layer of indirection?

Also, why is the order of the `fn` call and `Namespaces().Get()` reversed? And are the backoffs in the client not a good fit? You also construct a backoff per call, and elsewhere you have a package level backoff object.

smarterclayton replied:
This method takes a namespace and needs to return the last output namespace - that's different than what the client does by default.  We want to invoke fn() based on the first input, and then reload namespace later.

The backoffs in the client are not parameterized, so no.  It could be package level in this package, I guess.

krousey commented on pkg/controller/serviceaccount/tokens_controller.go:
```diff
@@ -31,12 +31,17 @@ import (
 )
 
-const NumServiceAccountRemoveReferenceRetries = 10
+var RemoveTokenBackoff = wait.Backoff{
```
Does this need to be a public package variable?

smarterclayton replied:
Yes, it's used in OpenShift.

krousey replied:
Since it's public, could you add a doc string?

krousey commented on pkg/controller/serviceaccount/tokens_controller.go:
```diff
@@ -242,16 +247,10 @@ func (e *TokensController) secretDeleted(obj interface{}) {
-			break
-		}
-		break
+	if err := client.RetryOnConflict(RemoveTokenBackoff, func() error {
```
Would one of the two client package's backoffs be fitting?

smarterclayton replied:
Possibly - @liggitt chose these backoffs based on empirical evidence though.

krousey commented:
The code mainly looks fine, but perhaps this shouldn't live in pkg/client/unversioned. There's nothing version specific about this, and I imagine you'd want to use this with the generated clients as well.

krousey commented:
@smarterclayton ping? Any thoughts on putting this somewhere else in the client directory?

smarterclayton commented:
Hrm - it really only needs to know about the client api error 409.  We
could put it in api errors, although that's a bit mucky.   Don't want to
create a utils package for it, for sure.

On Thu, Dec 17, 2015 at 1:03 PM, krousey <notifications@github.com> wrote:

> @smarterclayton <https://github.com/smarterclayton> ping? Any thoughts on
> putting this somewhere else in the client directory?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16067#issuecomment-165532955>
> .
>

smarterclayton commented:
Rebased this against master, dropped the namespace controller changes (didn't improve the code much) but now I want to implement this in https://github.com/kubernetes/kubernetes/issues/19848 in the repair controllers.  Addressed the remaining feedback, don't have a better place than conditions.go for now (happy to move conditions.go somewhere else if there is a suggestion), but would like to start using this.

ncdc commented:
@k8s-bot unit test this

smarterclayton commented:
ok to test

ncdc commented:
@k8s-bot e2e test this

smarterclayton commented:
Going to fix the retry now.

On Fri, Jan 22, 2016 at 1:23 PM, Zach Loafman <notifications@github.com>
wrote:

> Merged #16067 <https://github.com/kubernetes/kubernetes/pull/16067>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16067#event-523252785>.
>
